observer/holder/nodesHolder.go

```diff
@@ -0,0 +1,217 @@
+
+	// nodes not cached, compute the list and update the cache
+	recomputedList := make([]*data.NodeData, 0)
+	nh.mut.Lock()
+	for _, node := range nh.allNodes[shardID] {
+		if areCompatibleParameters(cache, node) {
+			recomputedList = append(recomputedList, node)
+		}
+	}
+	nh.cache[cacheKey] = recomputedList
+	nh.mut.Unlock()
+
+	return recomputedList
+}
+
+func areCompatibleParameters(cache cacheType, node *data.NodeData) bool {
+	isSynced, isFallback := node.IsSynced, node.IsFallback
+	if cache == syncedFallbackCache && isSynced && isFallback {
+		return true
+	}
+	if cache == outOfSyncFallbackNodes && !isSynced && isFallback {
+		return true
+	}
+	if cache == syncedNodesCache && isSynced && !isFallback {
+		return true
+	}
+	if cache == outOfSyncNodesCache && !isSynced && !isFallback {
+		return true
+	}
+
+	return false
+}
+
+func getCacheKey(cache cacheType, shardID uint32) string {
+	return fmt.Sprintf("%s_%d", cache, shardID)
+}
+
+// IsInterfaceNil returns true if there is no value under the interface
+func (nh *nodesHolder) IsInterfaceNil() bool {
+	return nh == nil
+}
+
+func (nh *nodesHolder) printNodesInShardsUnprotected() {
+	nodesByType := make(map[uint32]map[cacheType][]*data.NodeData)
+
+	// define a function to get the cache type for a node
+	getCacheType := func(node *data.NodeData) cacheType {
+		if node.IsFallback {
+			if node.IsSynced {
+				return syncedFallbackCache
+			}
+			return outOfSyncFallbackNodes
+		}
+		if node.IsSynced {
+			return syncedNodesCache
+		}
+		return outOfSyncNodesCache
+	}
+
+	// populate nodesByType map
+	for shard, nodes := range nh.allNodes {
+		if nodesByType[shard] == nil {
+			nodesByType[shard] = make(map[cacheType][]*data.NodeData)
+		}
+
+		for _, node := range nodes {
+			cache := getCacheType(node)
+			nodesByType[shard][cache] = append(nodesByType[shard][cache], node)
+		}
+	}
+
+	printHeader := "regular nodes"
+	if nh.availability == data.AvailabilityRecent {
+		printHeader = "snapshotless nodes"
+	}
+
+	for shard, nodesByCache := range nodesByType {
+		log.Info(fmt.Sprintf("shard %d %s", shard, printHeader),
+			"synced observers", getNodesListAsString(nodesByCache[syncedNodesCache]),
+			"synced fallback observers", getNodesListAsString(nodesByCache[syncedFallbackCache]),
+			"out of sync observers", getNodesListAsString(nodesByCache[outOfSyncNodesCache]),
+			"out of sync fallback observers", getNodesListAsString(nodesByCache[outOfSyncFallbackNodes]))
+	}
+}
+
+func getNodesListAsString(nodes []*data.NodeData) string {
+	addressesString := ""
+	for _, node := range nodes {
+		addressesString += fmt.Sprintf("%s, ", node.Address)
+	}
+
+	return strings.TrimSuffix(addressesString, ", ")
+}
+
+func computeInitialNodeList(regularNodes []*data.NodeData, fallbackNodes []*data.NodeData) map[uint32][]*data.NodeData {
+	mapToReturn := make(map[uint32][]*data.NodeData)
+	// in the first step, consider all the nodes to be active
+	for _, node := range regularNodes {
+		node.IsSynced = true
```

Comment:
is this ok? You change a property of the provided slice of nodes
Why not use a clone method?

Reply:
right. created a slice clone function and used it